### ISSUE TITLE
[css-text-3] Don't mince words about aliasing/shorthands with word-wrap

### ISSUE THREAD

**Issue by Manishearth** [w3c/csswg-drafts#866]
https://drafts.csswg.org/css-text-3/#propdef-word-wrap

The spec says:

> For legacy reasons, UAs must treat word-wrap as an alternate name for the overflow-wrap property, as if it were a shorthand of overflow-wrap.

The "as if it were a shorthand" is a bit hand-wavy, especially since "as an alternate name" can mean something else.

In particular, we have two different ways of "aliasing" properties. One is by making them shorthands of the aliasee. The other is by making them redirect to the other property at parse time.

The difference crops up when doing things like serializing property declaration blocks. For example, the following CSS:

```css
    #foo {
        word-wrap: break-word;
        overflow-wrap: break-word;
        align-items: flex-end;
        -webkit-align-items: baseline;
        foobar: baz;
    }
```

when serialized via `document.styleSheets[0].cssRules[0].cssText` in Firefox produces `"#foo { overflow-wrap: break-word; align-items: baseline; }"`, whereas in Chrome/Safari it produces `"#foo { word-wrap: break-word; overflow-wrap: break-word; align-items: baseline; }"`. This means that Firefox implements it as a "redirect alias", but Chrome/Safari implement it as a shorthand alias. All of them handle `-webkit-align-items` as a "redirect alias".

Browsers already use "redirect aliases" consistently for aliasing legacy prefixed properties. This concept of an alias is not defined in the spec, but is consistently implemented in browsers as something akin to:

If `foo` is an alias for property `bar`:

 - At parse time, if you encounter a property of name `foo`, treat it as if it were the name you encountered was of property `bar`
 - In CSSStyleDeclaration add accessors for the alias `foo` that under the hood work with property `bar`
 - Nowhere else should there be traces of the name `foo` (including during serialization of a declaration block)

The [compat spec _uses_ this concept of "aliasing"](https://compat.spec.whatwg.org/#css-simple-aliases), but does not define it.

We should either explicitly spec `word-wrap` to be a shorthand, or define this other concept of property alias in the spec and use it here (and also use it in the compat spec). I personally prefer the latter solution since aliasing is currently unspecced but imo shouldn't be.

cc @dholbert @dbaron @fantasai

**Comment by dholbert** [comment]
For reference, [here's a link to where Firefox implemented this property as an alias](https://bugzilla.mozilla.org/show_bug.cgi?id=955857) (rather than a shorthand).  I think we could just as easily implement it as a shorthand if it were clearly specced as such, though I agree with @Manishearth that an alias seems strictly simpler & more appropriate for legacy stuff.

**Comment by dbaron** [comment]
I'd also note that the normal rules for shorthands would produce a third result that is both (a) undesirable and (b) neither the Firefox nor the Chrome result, i.e., serializing the above rule to: <code>"#foo { word-wrap: break-word; align-items: baseline; }"</code>, since it's possible to "condense" the declaration into a shorthand.

This is by the same principle that:
```js
document.body.style.borderLeftColor = "green";
document.body.style.borderLeftWidth = "medium";
document.body.style.borderLeftStyle = "dotted";
console.log(document.body.getAttribute('style'));
```
produces:
```
border-left: medium dotted green;
```
in both Firefox and Chrome.

I think that also suggests that we need to specify the concept of aliases.

**Comment by dholbert** [comment]
FWIW, the spec-text for `word-wrap` used to be as follows:
> For legacy reasons, UAs may also accept ‘word-wrap’ as an alternate name for the ‘overflow-wrap’ property. However this syntax non-conforming in author style sheets.

https://www.w3.org/TR/2012/WD-css3-text-20120119/#overflow-wrap

The language was changed to its current form ("as an alternate name...as if it were a shorthand") in this commit in 2012: https://github.com/w3c/csswg-drafts/commit/05327146c505e17ec760efc6068c4864dba284ea

**Comment by dholbert** [comment]
(For the record, I filed https://bugs.chromium.org/p/chromium/issues/detail?id=679068 on the odd Blink/WebKit behavior here. It doesn't seem like they're actually treating it as a shorthand -- they're doing some special alternate form of aliasing, different from their standard legacy aliasing behavior.)

**Comment by Manishearth** [comment]
cc @miketaylr since the compat spec references this concept of aliasing without defining it (and we should take input from them if we start building the exact definition)

**Comment by frivoal** [comment]
The wording is unintentionally loose, but the intent was to make it a shorthand. If we want to keep that, we should update the phrasing to:

> For legacy reasons, UAs must treat word-wrap as an alternate name for the overflow-wrap property by implementing it as a shorthand of overflow-wrap.

Some of the reasoning for that is documented here: https://wiki.csswg.org/ideas/aliasing

TL;DR: we had a reasonably extensive discussion about how aliasing should work when "aliasing" break-before and page-break-before, and figured that shorthands were a good way to do that: well defined (including the OM) without needing to introduce any new concept, generally well behaved, deals well with the case when the legacy prop and the new prop don't have the exact same set of values.

I still think the idea has merit to it, but if implementations generally agree on a different aliasing mechanism, we should spec it and use that. Spec-fiction doesn't help anyone. Maybe we can keep exceptions and use shorthands when the legacy prop and the new prop accept different values.

**Comment by dholbert** [comment]
> I'd also note that the normal rules for shorthands would produce a third result that is...undesirable
> [serializing the modern property as its "shorthand" alias]

This doesn't actually seem to be what Chromium does for the few properties where they (ab)use shorthands to do aliasing.

Sample ([live jsfiddle version](https://jsfiddle.net/rfr5vgd1/)):
````
document.body.style.webkitColumnBreakInside = "avoid";
console.log(document.body.getAttribute('style'));
````
This testcase produces `break-inside: avoid;` even though they implement `-webkit-column-break-inside` as a shorthand for `break-inside`, as shown [here in their source](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/css/CSSProperties.json5?l=4467&rcl=8dfdd5c89b113097bbc34690f042cec6b4675363) and as noted in https://bugs.chromium.org/p/chromium/issues/detail?id=679068#c11

So, they do the Right Thing here, even though it's the opposite of normal shorthand behavior. (I wonder if that's due to a special-case codepath for shorthands that are only shorthands of a single property, maybe?)

**Comment by frivoal** [comment]
@dholbert @dbaron Since you consider shorthands as a bad aliasing mechanism, could you define what you consider to be a good one?

**Comment by csnardi** [comment]
Did anyone volunteer to write the text for the concept of an alias?

**Comment by frivoal** [comment]
Not as far as I can tell. I'm happy to update definitions to link to it once it's written, but someone with detailed knowledge of how it is implemented should write it down first.

**Comment by tabatkins** [comment]
Defined the CSS side of things at <https://drafts.csswg.org/css-cascade-4/#aliasing>. We think that's sufficient detail for Cascade; the remaining detail for how legacy shorthands work needs to show up in CSSOM, specifically in <https://drafts.csswg.org/cssom/#serializing-css-values> (or more precisely, probably in the definition of "preferred order").

**Comment by frivoal** [comment]
That looks like a good start.

However:
* For the alias part, it isn't clear to me that the spec text defines what's supposed to happen in APIs that return the name of the property, such as getting the value of the `transition-property` property. Maybe that's covered under "[...] This conversion also applies in the CSSOM, [...] for [...] property lookups", but I am not 100% sure. (the legacy shorthand part should be fine, since it is defined to behave like shorthands)
* I am a bit concerned that this is too vague though:
  > They otherwise behave exactly as regular shorthands, except that the CSSOM will not use them when serializing declarations. [CSSOM]

  To avoid having that statement being misinterpreted to applied to more things that we indeed, should we instead list the specific CSSOM APIs which are affected?

**Comment by tabatkins** [comment]
> For the alias part, it isn't clear to me that the spec text defines what's supposed to happen in APIs that return the name of the property, such as getting the value of the transition-property property. 

Hm, indeed. We should make that clear.

> To avoid having that statement being misinterpreted to applied to more things that we indeed, should we instead list the specific CSSOM APIs which are affected?

That's precisely the "remaining detail" that "needs to show up in CSSOM" I was talking about. No need for Cascade to list what needs to be changed in OM when OM can just change.

**Comment by fantasai** [comment]
> > For the alias part, it isn't clear to me that the spec text defines what's supposed to happen in APIs that return the name of the property, such as getting the value of the transition-property property. 
> 
> Hm, indeed. We should make that clear.

It is clear. The spec says

> When the old property’s syntax is identical to or a subset of the value space of the new property’s syntax, the two names are aliased with an operation on par with case-mapping: at parse time, the old property is converted into the new property. This conversion also applies in the CSSOM, both for string arguments and property lookups: requests for the old property transparently transfer to the new property instead.

Whatever is supposed to happen with a property "foo" inputted as "FOO", exactly the same happens between "new" and "old". If it is unclear what happens for case-mapping in some situation, then it is also unclear for aliasing; but once it is fixed for case-mapping it will be fixed for aliasing also.

**Comment by frivoal** [comment]
opened https://github.com/w3c/csswg-drafts/issues/3109 for the cssom part.

**Comment by frivoal** [comment]
We have a test for that: https://github.com/web-platform-tests/wpt/blob/master/css/css-text/overflow-wrap/word-wrap-alias.html, removing the label saying we need tests.
